### ISSUE TITLE
server: please log any call to StopWorkspace with the reason why it is called

### ISSUE THREAD

**Issue by sagor999** [gitpod-io/gitpod#12283]
## Is your feature request related to a problem? Please describe

On workspace side we only see `StopWorkspace` is being called. But looking at webapp logs, there is nothing in logs that would explain why StopWorkspace was called.

## Describe the behaviour you'd like

Can you please log a reason for anytime server is calling `StopWorkspace` to  ease with debug.
For example, if user requested to stop, or if some internal server logic decided to call it (like cancel prebuild because there is a newer commit).

**Comment by kylos101** [comment]
@sagor999 wdyt of adding traces too?

We might even be able to ensure that this particular trace is sampled at 100%. @mads-hartmann is it possible for us to configure certain traces such that they are always sampled, instead of at the default (I think 5 or 10% is the default for traces).

**Comment by mads-hartmann** [comment]
Ideally you would add both a log line and attach the information to the relevant span as the telemetry have different strengths depending on what kinds of questions you're trying to answer. For "system auditing" like this where we want to know why the system took a specific action against a specific workspace, logs are the best approach. If on the other hand you want to look at it more holistically, e.g. get a sense of what the distribution of "reasons" for stopping workspaces are, and you want to compute error rates, filter by cluster, etc. etc. then the tooling we have for traces are much better; and in that kind of analysis you really don't care about individual workspaces or requests, but the system as a whole.

I tried to capture some of this in [When to use what telemetry](https://www.notion.so/gitpod/When-to-use-what-telemetry-d38caafd594a45a598ad220149797e7d) but I haven't revised it since it was originally written so please comment if things are not clear 🧡 

Ideally we wouldn't have to worry about logs vs traces as a span is really just a structured log event with a fixed schema that captures the hierarchy of operations. I will create an issue to create a PoC that would allow us to set an attribute on a span (e.g. audit_log) that ensure that such spans are always sampled. With that in place we would be able to use spans for audit like logs, but I'm not a sure it's the right approach as I still think log lines are required in order for our self-hosted customers to understand why the system is behaving the way it is ☺️

**Comment by easyCZ** [comment]
So taking a higher level view, there's no reason we wouldn't attach the Stop reason as a first class property in the request, but also into the underlying WorkpsaceInstance representation (in both WebApp side, but also `ws-manager` side). That way, the reason can be consumed by both sides (and presumably both have a strong interest in this for SLOs - customer triggered Stop is good, if the system triggers Stop for other reasons, that could be an indicator, or perhaps you ran out of credits and we're forcibly stopping your workspaces - both parties should know to take the appropriate action).

Something like 
```
{
  ...
  stopRequested: bool
  stopReason: System readable reason
  stopDescription: Human readable reason (ie. You've run out of credits, please go and purchase more..)
}
```
The benefit of this is that you make it part of your domain model, which significantly easies your ability to debug - and to surface the reasons to the user.

**Comment by sagor999** [comment]
@easyCZ great idea!

**Comment by jankeromnes** [comment]
For reference, here are all the locations I could find where we issue a `StopWorkspaceRequest`:

<details><summary>When stopping a workspace via the API</summary>

https://github.com/gitpod-io/gitpod/blob/b634fb3836fb1ac546ccfbfc74a6f7a0ea32febc/components/server/src/workspace/gitpod-server-impl.ts#L734-L754

</details>

<details><summary>When deleting a workspace via the API</summary>

https://github.com/gitpod-io/gitpod/blob/b634fb3836fb1ac546ccfbfc74a6f7a0ea32febc/components/server/src/workspace/gitpod-server-impl.ts#L827-L841

</details>

<details><summary>When cancelling a prebuild via the API</summary>

https://github.com/gitpod-io/gitpod/blob/b634fb3836fb1ac546ccfbfc74a6f7a0ea32febc/components/server/ee/src/workspace/gitpod-server-impl.ts#L2610-L2627

</details>

<details><summary>When an admin force-stops a workspace form the admin panel</summary>

https://github.com/gitpod-io/gitpod/blob/b634fb3836fb1ac546ccfbfc74a6f7a0ea32febc/components/server/ee/src/workspace/gitpod-server-impl.ts#L858-L869

</details>

<details><summary>When we start a prebuild on a branch, we cancel all currently running prebuilds for that branch</summary>

https://github.com/gitpod-io/gitpod/blob/51c0c6b849513ba88e104e156880ddc085e576c4/components/server/ee/src/prebuilds/prebuild-manager.ts#L63-L97

</details>

<details><summary>When an admin blocks a user, we stop all their workspaces</summary>

https://github.com/gitpod-io/gitpod/blob/b634fb3836fb1ac546ccfbfc74a6f7a0ea32febc/components/server/ee/src/workspace/gitpod-server-impl.ts#L610-L641

</details>

<details><summary>When deleting a user, we stop all their workspaces (and then delete them afterwards)</summary>

https://github.com/gitpod-io/gitpod/blob/79b75ab4d5d979c927f1a3a77a128636cabce6a3/components/server/src/user/user-deletion-service.ts#L40-L107

</details>

<details><summary>When a user runs out of credits (creditAlert) -- although I'm not sure if this is still used? 🤔</summary>

https://github.com/gitpod-io/gitpod/blob/b634fb3836fb1ac546ccfbfc74a6f7a0ea32febc/components/server/ee/src/workspace/gitpod-server-impl.ts#L222-L251

</details>

**Comment by jankeromnes** [comment]
Thanks for the discussion & great ideas here!

In https://github.com/gitpod-io/gitpod/pull/12906 I've added a `reason` field to every `StopWorkspaceRequest` we issue, of type `StopWorkspaceReason` (a new enum).

- Does this look useful?
- Are the specified reasons helpful?
- Should this `reason` also be logged and/or added to tracing?

**Comment by jankeromnes** [comment]
Thanks again for the `StopWorkspaceReason` idea @easyCZ!

I gave this a first shot in https://github.com/gitpod-io/gitpod/pull/12906, and started with the following reasons:


```ts
enum StopWorkspaceReason {
    // The user stopped the workspace
    USER_STOPPED = 0;

    // The user deleted the workspace
    USER_DELETED = 1;

    // The user cancelled the running prebuild
    USER_CANCELLED_PREBUILD = 2;

    // An admin force-stopped the workspace
    ADMIN_STOPPED = 3;

    // An admin blocked the user
    ADMIN_BLOCKED = 4;

    // A new commit was pushed to a repository branch, so we cancelled all previously running prebuilds for that branch
    SYSTEM_CANCELLED_BRANCH_PREBUILD = 5;

    // The user exceeded their allowed usage limit (e.g. number of hours in month), so we stopped the workspace
    SYSTEM_USER_LIMIT_REACHED = 6;
}
```

However, after challenging this a bit, it still feels underspecified, so more discussion & alignment is likely needed before we can actually make such a change to the high-stakes `StopWorkspace` protocol.

Meanwhile (and possibly in parallel to the discussion above), maybe let's re-focus on the actual problem at hand, which was a lack of logging/tracing to help debug why workspaces are being stopped.

For that I can easily add logs & traces in a separate PR, given that we already know all the call sites.
